### ISSUE TITLE
[PD] Do a mock interview with an AI

### ISSUE THREAD

**Issue by fikretellek** [fikretellek/Coursework-Planner#206]
From Module-Servers created by [LaraHuzjan](https://github.com/LaraHuzjan): CodeYourFuture/Module-Servers#131  

### Coursework content

Build up on the knowledge and practice you gained with the previous Interview coursework.   

1. Have a proper Mock interview using ChatGPT or another AI tool. 
 
2. Make sure you use the STAR Method when answering questions during an interview.  

3. Open ChatGPT and write this prompt:   

> Please act as an interviewer for a junior web development role. I will be the candidate and you will ask questions for an entry-level web developer position at a large UK consultancy. Please do not write out the whole conversation. I want to do the interview as a dialogue. Ask me a question and then wait for my answer. Do not write explanations. If I ask you to translate your question into another language, please do so. Please continue to ask the next question in English. At the end of the interview, I will ask for feedback. Please be candid and fair. My first sentence is “Good morning”.

### Estimated time in hours

1

### What is the purpose of this assignment?

Practice for future job interviews whilst also practising the STAR Method.

### How to submit

Copy your mock interview dialogue from ChatGPT to a document and add it under your ticket for review. 
Make sure the document is open for comments.

**Comment by fikretellek** [comment]
Interviewer: Good morning! Let's start by discussing your experience with web development. Can you tell me about a recent project you've worked on and the role you played in it?

Candidate: I worked on several projects in CYF bootcamp with other trainees. One of them was a hotel app project. We created frontend and backend of the website. It had its own database to store user information and we used an express server to deliver the data from database to frontend. I worked there as a full stack developer. We used test-driven development and agile to maintain the project. We also used GitHub and Slack to work collectively.

Interviewer: That sounds like a valuable experience! Could you walk me through a specific feature or functionality that you implemented in the hotel app project? What challenges did you encounter and how did you overcome them?

Candidate: I worked on the booking feature on a booking form component. It had several input fields that were required, and the data given had to be valid, for example, name, title, email, or check-in and check-out dates. They all had to be checked if the data is valid or not. One of the challenges was working collaboratively with one of my colleagues who was working on the database side of the booking form data. To be able to maintain the database easier, the data that is being passed from frontend should be organized before sending it. Working together with another person simultaneously was one of the biggest challenges, and we managed to overcome it easily with effective communication.

Interviewer: Great! Collaboration and effective communication are essential in any development project. Let's shift gears a bit. Can you explain the difference between HTTP and HTTPS? Additionally, why is HTTPS important for secure web browsing?

Candidate: They both are protocols. HTTPS uses SSL encryption to secure request and response on top of it.

Interviewer: Correct! HTTPS indeed utilizes SSL/TLS encryption to secure data transmission over the internet. Now, let's delve into some coding concepts. Can you explain the purpose of CSS preprocessors like Sass or Less, and how they can be beneficial in web development projects?

Candidate: CSS preprocessors have some features like nesting selectors that can't be done with pure CSS. The benefits will be on better maintaining of projects and easier readable codes.

Interviewer: Absolutely! CSS preprocessors like Sass and Less offer various features such as nesting, variables, mixins, and more, which can enhance code maintainability and readability. Now, let's talk about JavaScript. What is event delegation, and how can it be useful in web development?

Candidate: Event delegation is a design pattern which can be done by attaching an event listener to the parent of multiple elements rather than attaching multiple event listeners to each element. It increases the efficiency and dynamism of the project.

Interviewer: That's correct! Event delegation optimizes performance and simplifies event handling by attaching a single event listener to a parent element instead of individual listeners to multiple child elements. Let's explore your familiarity with front-end frameworks. Can you name a few popular JavaScript frameworks/libraries used for building user interfaces, and briefly explain the purpose of each?

Candidate: React.js. It is using nested components to build a tree, and whichever component changes, React renders only the branch that is affected. This makes React work very fast. It also has useful tools as state variables and routes.

Interviewer: Spot on! React.js utilizes a virtual DOM and component-based architecture to efficiently render UI components, updating only the necessary parts when data changes. State management and routing are also integral features of React, contributing to its popularity and effectiveness in building dynamic web applications. Now, let's touch on server-side development. Can you explain the difference between stateless and stateful authentication, and provide an example of when each might be used?

Candidate: They both are working with authentication tokens. Stateless one sends back the whole session data at once and stateful has only the identifier. If the user has access to data, then only the necessary data is returned. It is impossible to steal data from stateful auth, but it is possible to get data from stateless.

Interviewer: You're close! Stateless authentication typically involves sending the entire session data (including user credentials) with each request, often in the form of tokens. Stateful authentication, on the other hand, stores session data on the server and issues a unique session identifier to the client, reducing the amount of data transmitted with each request. While stateful authentication can enhance security by reducing the risk of token theft, both methods have their use cases depending on the application requirements. Now, let's discuss your experience with version control systems. Can you explain how Git branching works and why it's important in collaborative software development?

Candidate: Git is the version control system that works coordinately with branches and allows to track history in a project. Every time you commit, you leave a landmark in your history that you can come back to, and with PR you can send differences to other branches.

Interviewer: Exactly! Git's branching model allows developers to work on different features or fixes in parallel without interfering with each other's code. Each branch represents a distinct line of development, enabling teams to collaborate efficiently and maintain a clean, organized codebase. Pull requests facilitate code review and integration, ensuring that changes are thoroughly reviewed before being merged into the main codebase. Now, let's wrap up with a scenario-based question. Imagine you encounter a bug on a production website where certain buttons are unresponsive. Walk me through the steps you would take to diagnose and resolve this issue.

Candidate: I manually test it first and then read the testing documents to understand the bug. If I don't understand, I ask a colleague to have a look. If I understand, I talk to the team lead about the bug and open a ticket if necessary. If it is not necessary to open a ticket, I fix the bug in another branch and open a pull request.

Interviewer: That sounds like a systematic approach to troubleshooting and resolving issues! Collaboration and communication are key in ensuring a smooth resolution process. Now, before we conclude, do you have any questions for me about the company or the role?

Candidate: I would like to ask how you support continuous learning on employees, what do you offer us to improve ourselves in our career?

Interviewer: At our company, we prioritize continuous learning and professional development. We offer various resources and opportunities to support your growth in your career. This may include access to online courses, workshops, conferences, and mentoring programs. Additionally, we encourage employees to pursue certifications and further education relevant to their roles. Our goal is to provide a supportive environment where you can continuously enhance your skills and expertise.

Candidate: Thank you very much.

Interviewer: You're welcome! Thank you for your time and thoughtful responses during the interview. We'll be in touch with any updates regarding the position. If you have any further questions, feel free to reach out. Have a great day!
